app/vmauth: authenticate by jwt token (#10435)
VictoriaMetrics/VictoriaMetrics: docs/victoriametrics/vmauth.md

See also [authorization](#authorization), [routing](#routing) and [load balancing](#load-balancing) docs.

> ### JWT Token auth proxy

Author: @TomFern I'd appreciate your review of the documentation

> **Warning**: JWT token authentication is still under development and may change in future releases. While we will try to minimize breaking changes, they may still occur. 

`vmauth` can authorize access to backends depending on the provided [JWT token](https://www.jwt.io/) in `Bearer Token` request headers. 
JWT tokens are verified using RSA or ECDSA public keys. The following auth config proxies requests to [single-node VictoriaMetrics](https://docs.victoriametrics.com/victoriametrics/single-server-victoriametrics/) if they contain a valid JWT token:

```yaml
users:
- jwt_token:
    public_keys:
    - |
      -----BEGIN PUBLIC KEY-----
      MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA...
      -----END PUBLIC KEY-----
  url_prefix: "http://victoria-metrics:8428/"
```

> JWT tokens must contain `vm_access` claim with tenant information:

Reviewer: ```suggestion
JWT tokens must contain a `vm_access` claim with tenant information:
```

```json
{
  "vm_access": {
    "tenant_id": {
      "account_id": 0,
      "project_id": 0
    }
  }
}
```

The claim `"vm_access": {}` is also allowed and means access to Tenant `0:0` will be provided. 
Optionally, vm_access claim might define `extra_labels` and `extra_filters` field to restrict access to only sub set of metrics.
The `vm_access` claim should work with `vmauth` and [vmgateway](https://docs.victoriametrics.com/victoriametrics/vmgateway/#access-control). 

For example, the following claim allows access to Tenant `0:0` with `env=prod` label and `job=foo` filter:
```json
{
  "vm_access": {
    "tenant_id": {
      "account_id": 0,
      "project_id": 0
    },
    "extra_labels": {
      "env": "prod"
    },
    "extra_filters": [
      "job=foo"
    ]
  }
}
```

The `url_prefix` supports placeholders for dynamic routing:
- `{{.Tenant}}` - replaced with `account_id:project_id` from token
- `{{.ExtraLabels}}` - replaced with metric labels from token
- `{{.ExtraFilters}}` - replaced with query filters from token

Foe example:
```
http://vmselect:8481/select/{{.Tenant}}/?extra_label={{.ExtraLabels}}&extra_filters={{.ExtraFilters}}
```

> For testing purposes, signature verification can be skipped via `skip_verify: true` (not recommended for production):

Reviewer: ```suggestion
For testing, skip signature verification with `skip_verify: true` (not recommended for production).
```
